File: src/dhlottery/pension-reserve.test.ts
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { PENSION_RESERVE_COST } from '../constants';
import type { HttpClient } from '../types';
import { DHLotteryError } from '../utils/errors';
import { reservePensionNextWeek } from './pension-reserve';

vi.mock('../notify/telegram', () => ({
  sendNotification: vi.fn().mockResolvedValue(undefined),
}));

vi.mock('./pension-crypto', () => ({
  encryptElQ: vi.fn((plain: string) => `enc:${plain}`),
  decryptElQ: vi.fn((encrypted: string) =>
    encrypted.startsWith('enc:') ? encrypted.slice(4) : encrypted
  ),
}));

const { sendNotification } = await import('../notify/telegram');

function createMockResponse(status: number, jsonData: unknown) {
  return {
    status,
    statusText: status === 200 ? 'OK' : 'ERROR',
    headers: new Headers(),
    text: async () => JSON.stringify(jsonData),
    json: async <T>() => jsonData as T,
  };
}

describe('pension reserve', () => {
  let mockClient: HttpClient;
  let fetchMock: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    fetchMock = vi.fn();
    mockClient = {
      cookies: {
        JSESSIONID: '12345678901234567890123456789012.tail.of.session==',
      },
      fetch: fetchMock,
      getCookieHeader: vi.fn(),
      clearCookies: vi.fn(),
    } as unknown as HttpClient;
    vi.clearAllMocks();
  });

  it('should reserve next week ticket and send success notification', async () => {
    fetchMock
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(
        createMockResponse(200, { resultCode: '100', resultMsg: 'ok', ROUND: '303', DRAW_DATE: '2026-02-19' })
      )
      .mockResolvedValueOnce(
        createMockResponse(200, { q: 'enc:{"resultCode":"100","resultMsg":"조회 성공","deposit":"10000"}' })
      )
      .mockResolvedValueOnce(
        createMockResponse(200, { q: 'enc:{"resultCode":"100","resultMsg":"예약 조회 성공","doubleRound":[]}' })
      )
      .mockResolvedValueOnce(
        createMockResponse(
          200,
          {
            q: 'enc:{"resultCode":"100","resultMsg":"예약 성공","reserveOrderNo":"A-1","reserveOrderDate":"2026-02-18 17:00:00"}',
          }
        )
      );

    const result = await reservePensionNextWeek(mockClient);

    expect(result).toMatchObject({
      status: 'success',
      success: true,
      skipped: false,
      targetRound: 304,
      totalAmount: PENSION_RESERVE_COST,
    });
    expect(fetchMock).toHaveBeenCalledTimes(6);
    expect(fetchMock.mock.calls[0][0]).toContain('/game/TotalGame.jsp?LottoId=LP72');
    expect(fetchMock.mock.calls[1][0]).toContain('/game/pension720/reserveGame.jsp');
    expect(fetchMock.mock.calls[2][0]).toContain('/roundRemainTime.do');
    expect(fetchMock.mock.calls[3][0]).toContain('/checkDeposit.do');
    expect(fetchMock.mock.calls[4][0]).toContain('/checkMyReserve.do');
    expect(fetchMock.mock.calls[5][0]).toContain('/addMyReserve.do');

    const duplicateBody = fetchMock.mock.calls[4][1]?.body as string;
    const duplicateQ = new URLSearchParams(duplicateBody).get('q');
    expect(duplicateQ).toContain('repeatRoundCnt=1');
    expect(duplicateQ).toContain('nextRound=304');
    expect(duplicateQ).toContain('winDate=');

    const addBody = fetchMock.mock.calls[5][1]?.body as string;
    const addQ = new URLSearchParams(addBody).get('q');
    expect(addQ).toContain('winDate=2026.02.26');

    expect(sendNotification).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'success',
        title: 'Pension Reserve Completed',
      })
    );
  });

  it('should skip when duplicate reserve exists for target round', async () => {
    fetchMock
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(
        createMockResponse(200, { resultCode: '100', resultMsg: 'ok', ROUND: '303', DRAW_DATE: '2026-02-19' })
      )
      .mockResolvedValueOnce(
        createMockResponse(200, { q: 'enc:{"resultCode":"100","resultMsg":"조회 성공","deposit":"10000"}' })
      )
      .mockResolvedValueOnce(
        createMockResponse(
          200,
          { q: 'enc:{"resultCode":"100","resultMsg":"예약 조회 성공","doubleRound":[{"doubleRound":"304","doubleCnt":"5"}]}' }
        )
      );

    const result = await reservePensionNextWeek(mockClient);

    expect(result).toMatchObject({
      status: 'skipped',
      success: true,
      skipped: true,
      targetRound: 304,
    });
    expect(fetchMock).toHaveBeenCalledTimes(5);
    expect(sendNotification).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'warning',
        title: 'Pension Reserve Skipped',
      })
    );
  });

  it('should return failure and warning when deposit is insufficient', async () => {
    fetchMock
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(
        createMockResponse(200, { resultCode: '100', resultMsg: 'ok', ROUND: '303', DRAW_DATE: '2026-02-19' })
      )
      .mockResolvedValueOnce(
        createMockResponse(200, { q: 'enc:{"resultCode":"100","resultMsg":"조회 성공","deposit":"3000"}' })
      );

    const result = await reservePensionNextWeek(mockClient);

    expect(result).toMatchObject({
      status: 'failure',
      success: false,
      code: 'PENSION_INSUFFICIENT_DEPOSIT',
      targetRound: 304,
    });
    expect(fetchMock).toHaveBeenCalledTimes(4);
    expect(sendNotification).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'warning',
        title: 'Pension Reserve Skipped',
      })
    );
  });

  it('should return failure on E001 auth error from duplicate check', async () => {
    fetchMock
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(createMockResponse(200, {}))
      .mockResolvedValueOnce(
        createMockResponse(200, { resultCode: '100', resultMsg: 'ok', ROUND: '303', DRAW_DATE: '2026-02-19' })
      )
      .mockResolvedValueOnce(
        createMockResponse(200, { q: 'enc:{"resultCode":"100","resultMsg":"조회 성공","deposit":"10000"}' })
      )
      .mockResolvedValueOnce(
        createMockResponse(200, { q: 'enc:{"resultCode":"E001","resultMsg":"로그인후 이용하시기 바랍니다.","doubleRound":[]}' })
      );

    const result = await reservePensionNextWeek(mockClient);

    expect(result).toMatchObject({
      status: 'failure',
      success: false,
      code: 'E001',
      targetRound: 304,
    });
    expect(sendNotification).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'error',
        title: 'Pension Reserve Failed',
      })
    );
  });

  it('should include DHLotteryError code in failure when bootstrap throws', async () => {
    fetchMock.mockRejectedValueOnce(
      new DHLotteryError('EL session failed', 'PENSION_BOOTSTRAP_FAILED')
    );

    const result = await reservePensionNextWeek(mockClient);

    expect(result).toMatchObject({
      status: 'failure',
      success: false,
      code: 'PENSION_BOOTSTRAP_FAILED',
    });
    expect(sendNotification).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'error',
        details: expect.objectContaining({ 오류코드: 'PENSION_BOOTSTRAP_FAILED' }),
      })
    );
  });

  it('should use PENSION_UNEXPECTED_ERROR code when a generic error is thrown', async () => {
    fetchMock.mockRejectedValueOnce(new TypeError('fetch failed'));

    const result = await reservePensionNextWeek(mockClient);

    expect(result).toMatchObject({
      status: 'failure',
      success: false,
      code: 'PENSION_UNEXPECTED_ERROR',
    });
    expect(sendNotification).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'error',
        details: expect.objectContaining({ 오류코드: 'PENSION_UNEXPECTED_ERROR' }),
      })
    );
  });
});

File: internal/dhlottery/pension_reserve_test.go
package dhlottery

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"strings"
	"testing"

	"github.com/kadragon/dhlottery-worker/internal/dherr"
	"github.com/kadragon/dhlottery-worker/internal/httpclient"
	"github.com/kadragon/dhlottery-worker/internal/notify"
	"github.com/kadragon/dhlottery-worker/internal/testutil"
)

const pensionSession = "12345678901234567890123456789012.tail.of.session=="

// encResp builds an EL response whose q field is the real-encrypted canned JSON
// (the production code decrypts it with the matching session).
func encResp(t *testing.T, jsonStr string) testutil.StubResponse {
	t.Helper()
	q, err := EncryptElQ(jsonStr, pensionSession)
	if err != nil {
		t.Fatal(err)
	}
	body, _ := json.Marshal(map[string]string{"q": q})
	return testutil.StubResponse{Status: 200, Body: string(body)}
}

// recordDecrypted decrypts an incoming EL request body and stores the parsed
// plaintext form so the test can assert on it.
func recordDecrypted(t *testing.T, r testutil.RecordedRequest, key string, into map[string]url.Values) {
	t.Helper()
	body, err := url.ParseQuery(r.Body)
	if err != nil {
		t.Fatal(err)
	}
	plain, err := DecryptElQ(body.Get("q"), pensionSession)
	if err != nil {
		t.Fatalf("decrypt request q: %v", err)
	}
	form, err := url.ParseQuery(plain)
	if err != nil {
		t.Fatal(err)
	}
	into[key] = form
}

type pensionScenario struct {
	deposit  string
	dup      string
	add      string
	firstErr error
}

func pensionClient(t *testing.T, s pensionScenario) (*httpclient.Client, map[string]url.Values, *testutil.StubDoer) {
	decrypted := map[string]url.Values{}
	stub := &testutil.StubDoer{Handler: func(_ int, r testutil.RecordedRequest) (testutil.StubResponse, error) {
		switch {
		case strings.Contains(r.URL, "TotalGame.jsp"):
			if s.firstErr != nil {
				return testutil.StubResponse{}, s.firstErr
			}
			return testutil.JSON("{}"), nil
		case strings.Contains(r.URL, "reserveGame.jsp"):
			return testutil.JSON("{}"), nil
		case strings.Contains(r.URL, "roundRemainTime.do"):
			return testutil.JSON(`{"resultCode":"100","resultMsg":"ok","ROUND":"303","DRAW_DATE":"2026-02-19"}`), nil
		case strings.Contains(r.URL, "checkDeposit.do"):
			recordDecrypted(t, r, "checkDeposit", decrypted)
			return encResp(t, s.deposit), nil
		case strings.Contains(r.URL, "checkMyReserve.do"):
			recordDecrypted(t, r, "checkMyReserve", decrypted)
			return encResp(t, s.dup), nil
		case strings.Contains(r.URL, "addMyReserve.do"):
			recordDecrypted(t, r, "addMyReserve", decrypted)
			return encResp(t, s.add), nil
		}
		return testutil.StubResponse{}, fmt.Errorf("unexpected %s", r.URL)
	}}
	client := httpclient.NewWithDoer(stub)
	client.SetCookie("JSESSIONID", pensionSession)
	return client, decrypted, stub
}

func TestPensionReserveSuccess(t *testing.T) {
	client, decrypted, _ := pensionClient(t, pensionScenario{
		deposit: `{"resultCode":"100","resultMsg":"조회 성공","deposit":"10000"}`,
		dup:     `{"resultCode":"100","resultMsg":"예약 조회 성공","doubleRound":[]}`,
		add:     `{"resultCode":"100","resultMsg":"예약 성공","reserveOrderNo":"A-1","reserveOrderDate":"2026-02-18 17:00:00"}`,
	})
	col := &notify.Collector{}
	out := reservePensionNextWeek(client, col)

	if out.Status != "success" || !out.Success || out.TargetRound != 304 || out.TotalAmount != 5000 {
		t.Fatalf("outcome = %+v", out)
	}

	// Real-crypto round trip: the server-side decrypt of our request must show
	// the correct next round and win date.
	if decrypted["checkMyReserve"].Get("nextRound") != "304" ||
		decrypted["checkMyReserve"].Get("repeatRoundCnt") != "1" ||
		decrypted["checkMyReserve"].Get("winDate") != "" {
		t.Errorf("checkMyReserve form = %v", decrypted["checkMyReserve"])
	}
	if decrypted["addMyReserve"].Get("winDate") != "2026.02.26" {
		t.Errorf("addMyReserve winDate = %q, want 2026.02.26", decrypted["addMyReserve"].Get("winDate"))
	}

	p := col.Payloads()[0]
	if p.Type != notify.Success || p.Title != "연금복권 예약 완료" || detailValue(p, "예약번호") != "A-1" {
		t.Errorf("payload = %+v", p)
	}
}

func TestPensionReserveDuplicate(t *testing.T) {
	client, _, stub := pensionClient(t, pensionScenario{
		deposit: `{"resultCode":"100","deposit":"10000"}`,
		dup:     `{"resultCode":"100","doubleRound":[{"doubleRound":"304","doubleCnt":"5"}]}`,
	})
	col := &notify.Collector{}
	out := reservePensionNextWeek(client, col)

	if out.Status != "skipped" || !out.Success || !out.Skipped || out.TargetRound != 304 {
		t.Fatalf("outcome = %+v", out)
	}
	if _, called := findRequest(stub, "addMyReserve"); called {
		t.Error("addMyReserve should not be called when duplicate exists")
	}
	if p := col.Payloads()[0]; p.Type != notify.Warning || p.Title != "연금복권 예약 건너뜀" {
		t.Errorf("payload = %+v", p)
	}
}

func TestPensionReserveInsufficientDeposit(t *testing.T) {
	client, _, _ := pensionClient(t, pensionScenario{
		deposit: `{"resultCode":"100","deposit":"3000"}`,
	})
	col := &notify.Collector{}
	out := reservePensionNextWeek(client, col)

	if out.Status != "failure" || out.Code != "PENSION_INSUFFICIENT_DEPOSIT" || out.TargetRound != 304 {
		t.Fatalf("outcome = %+v", out)
	}
	if p := col.Payloads()[0]; p.Type != notify.Warning || p.Title != "연금복권 예약 건너뜀" {
		t.Errorf("payload = %+v", p)
	}
}

func TestPensionReserveDuplicateAuthError(t *testing.T) {
	client, _, _ := pensionClient(t, pensionScenario{
		deposit: `{"resultCode":"100","deposit":"10000"}`,
		dup:     `{"resultCode":"E001","resultMsg":"로그인후 이용하시기 바랍니다.","doubleRound":[]}`,
	})
	col := &notify.Collector{}
	out := reservePensionNextWeek(client, col)

	if out.Status != "failure" || out.Code != "E001" || out.TargetRound != 304 {
		t.Fatalf("outcome = %+v", out)
	}
	if p := col.Payloads()[0]; p.Type != notify.Error || p.Title != "연금복권 예약 실패" {
		t.Errorf("payload = %+v", p)
	}
}

func TestPensionReserveBootstrapError(t *testing.T) {
	client, _, _ := pensionClient(t, pensionScenario{
		firstErr: dherr.New("EL session failed", "PENSION_BOOTSTRAP_FAILED"),
	})
	col := &notify.Collector{}
	out := reservePensionNextWeek(client, col)

	if out.Status != "failure" || out.Code != "PENSION_BOOTSTRAP_FAILED" {
		t.Fatalf("outcome = %+v", out)
	}
	if p := col.Payloads()[0]; p.Type != notify.Error || detailValue(p, "오류코드") != "PENSION_BOOTSTRAP_FAILED" {
		t.Errorf("payload = %+v", p)
	}
}

func TestPensionReserveGenericError(t *testing.T) {
	client, _, _ := pensionClient(t, pensionScenario{
		firstErr: errors.New("fetch failed"),
	})
	col := &notify.Collector{}
	out := reservePensionNextWeek(client, col)

	if out.Status != "failure" || out.Code != "PENSION_UNEXPECTED_ERROR" {
		t.Fatalf("outcome = %+v", out)
	}
	if p := col.Payloads()[0]; p.Type != notify.Error || detailValue(p, "오류코드") != "PENSION_UNEXPECTED_ERROR" {
		t.Errorf("payload = %+v", p)
	}
}

func TestPensionReserveMissingSession(t *testing.T) {
	// No JSESSIONID/DHJSESSIONID cookie -> postEncrypted fails at getSessionID.
	stub := &testutil.StubDoer{Handler: func(_ int, r testutil.RecordedRequest) (testutil.StubResponse, error) {
		switch {
		case strings.Contains(r.URL, "TotalGame.jsp"), strings.Contains(r.URL, "reserveGame.jsp"):
			return testutil.JSON("{}"), nil
		case strings.Contains(r.URL, "roundRemainTime.do"):
			return testutil.JSON(`{"resultCode":"100","ROUND":"303","DRAW_DATE":"2026-02-19"}`), nil
		}
		return testutil.StubResponse{}, fmt.Errorf("unexpected %s", r.URL)
	}}
	client := httpclient.NewWithDoer(stub)
	col := &notify.Collector{}
	out := reservePensionNextWeek(client, col)
	if out.Status != "failure" || out.Code != "PENSION_AUTH_MISSING" {
		t.Errorf("outcome = %+v", out)
	}
}
